### PR TITLE
Issue #3508/ added token command for CRUD operations

### LINKED ISSUE #3508: [mesheryctl] CRUD for tokens
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior

Users can have multiple tokens and reference them in their meshconfig (~/.meshery/config.yaml) file. This means that a user can have multiple tokens and can use different tokens for different contexts.

However, right now the users have to manually edit their config files in-order to change the token used in a particular context and even add new tokens, change the location of the token etc.

#### Desired Behavior

Similar to channel and context constructs in the meshconfig, mesheryctl should also support CRUD actions for tokens:

Add a new token to the meshconfig:

* `mesheryctl system token add <token name> -f <optional: path to the token, defaults to auth.json>`

Delete token:

* `mesheryctl system token delete <token name>`

Set token (of current-context or context passed in with the `--context` flag):

* `mesheryctl system token set <token name>`

List tokens:

* `mesheryctl system token list`

View tokens:

* `mesheryctl system token view <token name>` (if token name is not passed, then the token of the current-context)

Power/advanced users can directly edit the meshconfig for use cases other than these.

---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by Revolyssup**
Signed-off-by: Ashish Tiwari <ashishjaitiwari15112000@gmail.com>

**Description**

This PR fixes #3508 

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by Revolyssup** [comment]
@navendu-pottekkat  Can you review it?

**Comment by Shreyas220** [comment]
@Revolyssup amazing work :smile: 
@piyushsingariya ,i need some help , i tested and all the commands are working, could tell us why the unit test is failing ?

**Comment by Revolyssup** [comment]
> @Revolyssup amazing work 
> @piyushsingariya ,i need some help , i tested and all the commands are working, could tell us why the unit test is failing ?

@Shreyas220  I checked in the actions workflow, it was some Codecov specific problem.
[Error: Codecov failed with the following error: The process '/usr/bin/bash' failed with exit code 1]

**Comment by Shreyas220** [comment]
> > @Revolyssup amazing work
> > @piyushsingariya ,i need some help , i tested and all the commands are working, could tell us why the unit test is failing ?
> 
> @Shreyas220 I checked in the actions workflow, it was some Codecov specific problem.
> [Error: Codecov failed with the following error: The process '/usr/bin/bash' failed with exit code 1]

yes @Revolyssup, i saw that , but i am afraid i dont understand the error :upside_down_face:, hence calling the expert

**Comment by Revolyssup** [comment]
> > > @Revolyssup amazing work
> > > @piyushsingariya ,i need some help , i tested and all the commands are working, could tell us why the unit test is failing ?
> > 
> > 
> > @Shreyas220 I checked in the actions workflow, it was some Codecov specific problem.
> > [Error: Codecov failed with the following error: The process '/usr/bin/bash' failed with exit code 1]
> 
> yes @Revolyssup, i saw that , but i am afraid i dont understand the error , hence calling the expert

@Shreyas220  Should I create another commit. I can add Long description for add and delete commands which aren't there right now. Printing their usage like, "mesheryctl system token add <token name>"

File: mesheryctl/internal/cli/root/system/token.go [hunk 1]
```diff
@@ -30,7 +30,12 @@ var tokenCmd = &cobra.Command{
 var addTokenCmd = &cobra.Command{
 	Use:   "add",
 	Short: "Add a token to config.yaml",
-	Args:  cobra.ExactArgs(1),
+	Long: `
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/system/token.go:33]:
```suggestion
	Example: `
```
"Long" is the longer version of the description.

**Review by piyushsingariya — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/token.go (hunk 1 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:33]: okay,will do the change

Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go]: @piyushsingariya  did the change

File: mesheryctl/internal/cli/root/system/token.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var ctx string
+var tokenCmd = &cobra.Command{
+	Use:   "token",
+	Short: "Perform CRUD operations on token",
+	Long: `
+	Add, Delete and Modify tokens in config.yaml`,
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go]: @leecalcote  made the changes.

**Comment by leecalcote** [comment]
@Revolyssup lets plan on a demo from you on Wednesday.

**Comment by leecalcote** [comment]
Have you considered how this integrates with `system login`?

**Comment by leecalcote** [comment]
Do we have unit tests included for common pitfalls?

File: mesheryctl/internal/cli/root/system/token.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/token.go]: Good man. Your welcome to resolve any of my comments that you’ve fully addressed.

**Comment by Revolyssup** [comment]
> Have you considered how this integrates with `system login`?

@leecalcote  system login currently looks for token name inside current context and then add that in the tokens array along with the location where it stored the token.
So user will have to set the name for token with "mesheryctl system token set <token name>" and then use system login. Here if the token field is unspecified/empty in current context, an entry with empty token name is added in tokens array.
So we can add an additional check in login.go . And if the token.Name is empty, we can throw an error that "No token name set, set a token name before logging in". Problem with automatically adding Default name is that, a token with that name might already exist in the tokens array. You might have a better idea for such case.

**Comment by Revolyssup** [comment]
> Do we have unit tests included for common pitfalls?

The problem with unit tests that we don't have any or too few right now. Piyush is slowly creating issue and as this is an epic issue, he is assigning unit tests for different commands slowly. For the same reason my last PR for enhancing view with token-location is pending because we dont have unit tests. If you want I can write unit test for token.go covering all different edge cases.

**Comment by Revolyssup** [comment]
> @Revolyssup lets plan on a demo from you on Wednesday.

Yups sure!

**Comment by leecalcote** [comment]
> > Do we have unit tests included for common pitfalls?
> 
> The problem with unit tests is that we don't have any or too few right now. Piyush is slowly creating issue and as this is an epic issue, he is assigning unit tests for different commands slowly. For the same reason my last PR for enhancing view with token-location is pending because we dont have unit tests. If you want I can write unit test for token.go covering all different edge cases.

Unless there's a specific convention that @piyushsingariya, @navendu-pottekkat and @hexxdump are setting forth to be followed, yes, please plaster unit tests all over this sucker.

File: mesheryctl/internal/cli/root/system/token.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if ok := utils.IsValidSubcommand(availableSubcommands, args[0]); !ok {
+			return errors.New(utils.SystemError(fmt.Sprintf("invalid command: \"%s\"", args[0])))
+		}
+		return nil
+	},
+}
+
+var addTokenCmd = &cobra.Command{
+	Use:   "add",
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:31]: @navendu-pottekkat we’re using add here and not create like we do for context?

File: mesheryctl/internal/cli/root/system/token.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	Short: "Add a token to meshery config",
+	Example: `
+	mesheryctl system token add <token-name> -f <token-path>
+	mesheryctl system token add <token-name> (default path is auth.json)
+	`,
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		tokenName := args[0]
+		if tokenPath == "" {
+			tokenPath = "auth.json"
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:41]: @utkarsh-pro @navendu-pottekkat we decided on the default path being the location of `.meshery` folder, correct? @Revolyssup, was this requirement specified in the issue?

File: mesheryctl/internal/cli/root/config/config.go [hunk 5]
```diff
@@ -231,6 +231,75 @@ func AddTokenToConfig(token Token, configPath string) error {
...
+
+	viper.SetConfigFile(configPath)
+	err := viper.ReadInConfig()
+	if err != nil {
+		return err
+	}
+
+	mctlCfg, err := GetMesheryCtl(viper.GetViper())
+	if err != nil {
+		return errors.Wrap(err, "error processing config")
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/config/config.go:248]: Let's use errors.new() - MeshKit style errors, yeah? 

@Shreyas220, be a stickler about this, please.

File: mesheryctl/internal/cli/root/system/token.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		}
+
+		token := config.Token{
+			Name:     tokenName,
+			Location: tokenPath,
+		}
+		if err := config.AddTokenToConfig(token, utils.DefaultConfigPath); err != nil {
+			return errors.Wrap(err, "Could not add specified token to config")
+		}
+		log.Printf("Token %s added successfully!", tokenName)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:51]:
```suggestion
		log.Printf("Token %s added.", tokenName)
```

File: mesheryctl/internal/cli/root/system/token.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+		return nil
+	},
+}
+var deleteTokenCmd = &cobra.Command{
+	Use:   "delete",
+	Short: "Delete a token from meshery config",
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:57]: Always capitalize Meshery.

Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:57]:
```suggestion
	Short: "Delete a token your meshconfig",
```

File: mesheryctl/internal/cli/root/system/token.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	Example: `
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:58]: Where is the long description?

File: mesheryctl/internal/cli/root/system/token.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	`,
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		tokenName := args[0]
+
+		token := config.Token{
+			Name:     tokenName,
+			Location: tokenPath,
+		}
+		if err := config.DeleteTokenFromConfig(token, utils.DefaultConfigPath); err != nil {
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:69]: Can we give the name of the token that couldn't be deleted?

File: mesheryctl/internal/cli/root/system/token.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+			return errors.Wrap(err, "Could not delete specified token from config")
+		}
+		log.Printf("Token %s deleted successfully!", tokenName)
+		return nil
+	},
+}
+var setTokenCmd = &cobra.Command{
+	Use:   "set",
+	Short: "Set token for context",
+	Long:  "Set Token for current context or context passed with --context flag.",
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:79]: Why capital Token in long descr and lowercase in short descr. Let's go with lowercase.

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/token.go:79]:
```suggestion
	Long:  "Set token for the current context or the context specified with the --context flag.",
```

File: mesheryctl/internal/cli/root/system/token.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	`,
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		tokenName := args[0]
+		if ctx == "" {
+			ctx = viper.GetString("current-context")
+
+		}
+		if err := config.SetTokenToConfig(tokenName, utils.DefaultConfigPath, ctx); err != nil {
+			return errors.Wrapf(err, "Could not set specified token on conteext %s", ctx)
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/token.go:92]:
```suggestion
			return errors.Wrapf(err, "Could not set specified token on context %s", ctx)
```

Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/token.go:92]: Can you provide the token name in this error?

**Comment by leecalcote** [comment]
Can we give the user a pick list of token names to chose from in the event that they do not provide a valid token name?

```
➜  mesheryctl git:(Revolyssup/feature/crudForTokens) ✗ ./mesheryctl system token view default
Error: Token default could not be found.
Usage:
  mesheryctl system token view [flags]

Examples:

	mesheryctl system token view <token-name>
	mesheryctl system token view (show token of current context)


Flags:
  -h, --help   help for view

Global Flags:
      --config string    config file (default location is: /Users/lee/.meshery/config.yaml) (default "/Users/lee/.meshery/config.yaml")
  -c, --context string   (optional) temporarily change the current context.
  -v, --verbose          verbose output
  -y, --yes              (optional) assume yes for user interactive prompts.
```

**Comment by leecalcote** [comment]
@Revolyssup it would be great to support `mesheryctl system token view --all` - the `--all` flag. We'll want to provide a common set of subcommands and flags across all `system` resources (and non-system like `pattern` and `filter`).

File: mesheryctl/internal/cli/root/system/token.go (hunk 4 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:41]: The issue asked to set the default path as auth.json.

File: mesheryctl/internal/cli/root/system/token.go (hunk 7 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:57]: Will remember.

File: mesheryctl/internal/cli/root/system/token.go (hunk 8 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:58]: Example and Short description pretty much summed up the command so didn't add Long. But to stick with convention, will add for sure now.

File: mesheryctl/internal/cli/root/system/token.go (hunk 9 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:69]: Made the change.

File: mesheryctl/internal/cli/root/system/token.go (hunk 11 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:92]: Yups

File: mesheryctl/internal/cli/root/system/token.go (hunk 10 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/token.go:79]: > @Revolyssup it would be great to support `mesheryctl system token view --all` - the `--all` flag. We'll want to provide a common set of subcommands and flags across all `system` resources (and non-system like `pattern` and `filter`).



> @Revolyssup it would be great to support `mesheryctl system token view --all` - the `--all` flag. We'll want to provide a common set of subcommands and flags across all `system` resources (and non-system like `pattern` and `filter`).

Yups, was actually adding it but it wasn't specified in issue. Added it now.

**Comment by Revolyssup** [comment]
> @Revolyssup it would be great to support `mesheryctl system token view --all` - the `--all` flag. We'll want to provide a common set of subcommands and flags across all `system` resources (and non-system like `pattern` and `filter`).

Added it.

File: mesheryctl/internal/cli/root/system/token.go (hunk 3 above)
Comment (piyushsingariya) [reply mesheryctl/internal/cli/root/system/token.go:31]: Yes, create is something that eventually means creation, which is not exactly the case with tokens, we aren't creating a token, we are adding it to the config. That is why @navendu-pottekkat replaced to with "add" instead of "create"

File: mesheryctl/internal/cli/root/system/token.go (hunk 4 above)
Comment (piyushsingariya) [reply mesheryctl/internal/cli/root/system/token.go:41]: Yes, the default path is ~/.meshery folder, but we eventually don't need to mention ~/.meshery in the location it can directly be auth.json we are handling it later with the help of few functions [https://github.com/meshery/meshery/blob/master/mesheryctl/internal/cli/root/constants/constants.go#L54](here)

**Comment by Revolyssup** [comment]
@leecalcote  Should I add --all flag for delete too? In this case, user might be prompted for confirmation, and if userResponse is true, we can delete all the tokens.

**Comment by Revolyssup** [comment]
> Can we give the user a pick list of token names to chose from in the event that they do not provide a valid token name?

`token list` command already does that. So what if along with the error, we prompt "Use `mesheryctl system token list` to see available tokens."

**Comment by leecalcote** [comment]
@Revolyssup FYI, if it saves you time, you can click the "Commit suggestion" button if you are accepted a suggested change verbatim.

**Comment by leecalcote** [comment]
> > Can we give the user a pick list of token names to chose from in the event that they do not provide a valid token name?
> 
> `token list` command already does that. So what if along with the error, we prompt "Use `mesheryctl system token list` to see available tokens."

I'm suggesting to do that for them using the `promptui` library. Example - https://github.com/meshery/meshery/pull/3514/files#diff-1959e4ba20d3a9031de02dcaf2cd159f2c6ad5c8e580d052d6a1268ea4689770R170

**Comment by Revolyssup** [comment]
> > > Can we give the user a pick list of token names to chose from in the event that they do not provide a valid token name?
> > 
> > 
> > `token list` command already does that. So what if along with the error, we prompt "Use `mesheryctl system token list` to see available tokens."
> 
> I'm suggesting to do that for them using the `promptui` library. Example - https://github.com/meshery/meshery/pull/3514/files#diff-1959e4ba20d3a9031de02dcaf2cd159f2c6ad5c8e580d052d6a1268ea4689770R170

Got it. Will do

**Comment by Revolyssup** [comment]
@leecalcote  If everything looks fine to you, then I can start writing unit tests for token command.

**Review by navendu-pottekkat — COMMENTED** [review]
@Revolyssup I think it is best if we hold this PR a little bit longer as I am making changes to how mesheryctl handles the config file. This would make reading and writing to meshconfig file easy and it will make it consistent. We would be able to use that changes here and we would be able to avoid a lot of technical debt. This issue has nothing to do with this PR but it is a part of the refactoring initiative I have taken to.

Rest of the CRUD operations looks good to me 💯

**Comment by leecalcote** [comment]
Hang in there, @Revolyssup. You're _this_ close to the finish line. :)

**Comment by Revolyssup** [comment]
@navendu-pottekkat  can you review and suggest changes?

File: mesheryctl/internal/cli/root/config/config.go [hunk 12]
```diff
@@ -322,6 +347,71 @@ func AddTokenToConfig(token Token, configPath string) error {
 	return nil
 }
 
+// DeleteTokenFromConfig deletes a token passed to it to mesheryctl config file
+func DeleteTokenFromConfig(tokenName string, configPath string) error {
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
```
Comment (hexxdump) [inline mesheryctl/internal/cli/root/config/config.go:352]: Can we write a generic function GetConfig(configPath string), which returns MesheryCtlConfig?
I see this code is common to all sub commands.

Comment (Revolyssup) [reply mesheryctl/internal/cli/root/config/config.go:352]: Initially I was doing almost something like that. But navendu asked me to avoid extra functions. So I removed those functions.

Comment (hexxdump) [reply mesheryctl/internal/cli/root/config/config.go:352]: @navendu-pottekkat, The code from line:352-365 has repeated 4 times in this file. It would be good to convert it into a function as per [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
But it would require some rework to AddTokenToConfig , DeleteTokenFromConfig and SetTokenToConfig functions. i.e. to pass MesheryCtlConfig directly instead of utils.DefaultConfigPath

Let me know what you think.

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/config/config.go:352]: Yes, ideally we need to call this only on a top level command. I am looking into how we can prevent reading in the config file for each of the subcommands. Instead, we read the config once and then can just reference the central struct in which it will be stored. I am currently working on this refactor and it would be great if we could all come back to this holistically. I recently merged a PR that centralises writes to config file for contexts, a similar implementation could be added here as well.

We can make the changes here as @hexxdump and @Revolyssup has suggested in a separate PR as it would complicate things otherwise. If the current changes are good functionally, let's merge this.

Comment (Revolyssup) [reply mesheryctl/internal/cli/root/config/config.go:352]: @navendu-pottekkat  sounds about right

Comment (Revolyssup) [reply mesheryctl/internal/cli/root/config/config.go:352]: @navendu-pottekkat  Even after future refactoring, functionally things would be same, so I will get going with tests for token command.

Comment (hexxdump) [reply mesheryctl/internal/cli/root/config/config.go:352]: Yes, this should be good for now. 
I also thought about reading the file once and caching the data for subsequent use, however was concerned about the set operations. We need to revalidate the cache in those scenarios.

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/config/config.go:352]: yes, will you be adding tests in this PR?

Comment (Revolyssup) [reply mesheryctl/internal/cli/root/config/config.go:352]: @navendu-pottekkat  I already created another branch for tests, thinking of keeping the PR's for tests separate.But I will proceed in whatever way you recommend.

Comment (navendu-pottekkat) [reply mesheryctl/internal/cli/root/config/config.go:352]: Okay, sounds good. Let's then get this merged.

**Review by navendu-pottekkat — APPROVED** [review]
